javatests/arcs/android/storage/service/StorageServiceNgImplTest.kt

```diff
@@ -0,0 +1,245 @@
+    assertThat(storageServiceNg.stores.size).isEqualTo(0)
+
+    // Define the store options for two different stores.
+    val storeOptions1 = StoreOptions(RamDiskStorageKey("store1"), CountType())
+    val encodedStoreOptions1 = with(Parcel.obtain()) {
+      writeStoreOptions(storeOptions1, ParcelableCrdtType.Count, 0)
+      marshall()
+    }
+
+    val storeOptions2 = StoreOptions(RamDiskStorageKey("store2"), CountType())
+    val encodedStoreOptions2 = with(Parcel.obtain()) {
+      writeStoreOptions(storeOptions2, ParcelableCrdtType.Count, 0)
+      marshall()
+    }
+
+    // Open a storage channel for one store and assert that a new store has been created.
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions1,
+      createChannelCallback(deferredChannel1),
+      messageCallback
+    )
+    val channel1 = deferredChannel1.await()
+    assertThat(storageServiceNg.stores.size).isEqualTo(1)
+
+    // Open a storage channel for another store and assert that another new store has been created.
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions2,
+      createChannelCallback(deferredChannel2),
+      messageCallback
+    )
+    val channel2 = deferredChannel2.await()
+    assertThat(storageServiceNg.stores.size).isEqualTo(2)
+
+    // Also check that the two channels that were created are distinct.
+    assertThat(channel2).isNotSameInstanceAs(channel1)
+  }
+
+  @Test
+  fun openingTwoChannels_forSameStore_onlyCreatesOneStore() = runBlockingTest {
+    val deferredChannel1 = CompletableDeferred<IStorageChannel>()
+    val deferredChannel2 = CompletableDeferred<IStorageChannel>()
+
+    val storageServiceNg = createStorageService(this) as StorageServiceNgImpl
+    assertThat(storageServiceNg.stores.size).isEqualTo(0)
+
+    // Define the store options for a store.
+    val storeOptions1 = StoreOptions(RamDiskStorageKey("store1"), CountType())
+    val encodedStoreOptions1 = with(Parcel.obtain()) {
+      writeStoreOptions(storeOptions1, ParcelableCrdtType.Count, 0)
+      marshall()
+    }
+
+    // Open a storage channel for the store and assert that a new store has been created.
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions1,
+      createChannelCallback(deferredChannel1),
+      messageCallback
+    )
+    val channel1 = deferredChannel1.await()
+    assertThat(storageServiceNg.stores.size).isEqualTo(1)
+
+    // Open another storage channel for the same store. Ensure no new store has been created.
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions1,
+      createChannelCallback(deferredChannel2),
+      messageCallback
+    )
+    val channel2 = deferredChannel2.await()
+    assertThat(storageServiceNg.stores.size).isEqualTo(1)
+
+    // Also check that the two channels that were created are distinct.
+    assertThat(channel2).isNotSameInstanceAs(channel1)
+  }
+
+  private fun createStorageService(scope: CoroutineScope): IStorageServiceNg {
```

Comment:
This can go in the `setUp` method, storing the result in a `storageService` member variable. It looks like the only reason you wanted this as a helper method was to pass in the `runBlockingTest` scope.

@jblebrun proposed a new pattern, which I don't believe anyone has actually employed yet, but should work, where you make `scope` a member variable as well, and instead of calling `runBlockingTest` directly, you call `scope.runBlockingTest` instead.

Have a read of the section titled "Providing an explicit TestCoroutineScope" from https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/

The pattern looks roughly like this:

```
private val scope = TestCoroutineScope()
private lateinit var storageService: StorageServiceNgImpl

@Before
fun setUp() {
  storageService = StorageServiceNgImpl(...)
}

@After
fun tearDown() {
  scope.cleanupTestCoroutines()
}

@Test
fun testFoo() = scope.runBlockingTest {
  storageService.whatever(scope)
}
```

Reply:
+100

Reply:
done :)